Templates vignette (#122)
GSK-Biostatistics/tfrmt: vignettes/layer.Rmd

)
```

```{r setup, message=FALSE}
library(tfrmt)
library(dplyr)
```

A core design element of `tfrmt` is to offer the ability to layer tfrmts together. This ability provides an opportunity to build template table formats that can be shared, improved, and reused across multiple projects. `tfrmt` provides a few tfrmt templates to facilitate the creation of basic tfrmts, but they require customization. This can be done through layering. 

To layer tfrmts, you can pipe or use the `layer_tfrmt()` function. Piping is the preferred method for readability.

> When a tfrmt gets layered, the values in the first tfrmt are replaced with the values in the second tfrmt. The exception to this rule is the "body_plan", which will stack the plans together. To change this behavior see the `layer_tfrmt()` documentation.

Reviewer: ```suggestion
This layering function coalesces the values from the input tfrmts to form a new tfrmt. If the input tfrmts have any of the same parameters specified, the layering function will prioritize those provided in the latter. The exception to this rule is the "body_plan", which will stack the plans together. To change this behavior see the `layer_tfrmt()` documentation.
```

## Example

> Here is an example of laying through piping. We create a template from `tfrmt_sigdig()` and pipe it into the `tfrmt`. 

Reviewer: ```suggestion
Here is an example of layering through piping. We create a template from `tfrmt_sigdig()` and pipe it into the `tfrmt`. 
```

```{r setup}
library(tlang)
library(magrittr)
```

> A core design element of `tlang` was to offer the ability to layer and build template table formats that could be shared, improved, and eventually customized for a specific table. The idea being that organizations could develop a standard `tfrmt` object that would serve as the base format for the rest of the tables, and make it so programmers could focus on setting up the data as opposed to formatting of the table. 

Reviewer: ```suggestion
A core design element of `tlang` is to offer the ability to layer and build template table formats that can be shared, improved, and eventually customized for a specific table. The idea being that organizations can develop a standard `tfrmt` object that can be reused across many different tables, and make it so programmers can focus on setting up the data as opposed to formatting of the table. 
```

> There are two ways that `tfrmt` objects may be layered. The preferred method is through the first argument of `tfrmt` - `tfrmt_obj`. The second, but perhaps more explicit method is using the `layer_tfrmt` function. We will go through their applications below.  

Reviewer: It might be good to describe these in more laymans terms and then use the same  description in the rest of the doc.

Reviewer 2: I agree I think the information about layer_tfrmt needs to be moved out of here and into the help docs for layer_tfrmt

Reviewer 2: I agree I think the information about layer_tfrmt needs to be moved out of here and into the help docs for layer_tfrmt

> There are two ways that `tfrmt` objects may be layered. The preferred method is through the first argument of `tfrmt` - `tfrmt_obj`. The second, but perhaps more explicit method is using the `layer_tfrmt` function. We will go through their application below.  

Reviewer: ```suggestion
There are two ways that `tfrmt` objects may be layered. The preferred method is through the first argument of `tfrmt` - `tfrmt_obj`. The second, but perhaps more explicit method is using the `layer_tfrmt` function. We will go through their applications below.  
```

## Creating a template

Before we show how to layer a tfrmt, we need to create a template tfrmt. A template is intended to provide the basics values that an organization would expect to be repeated across multiple tables. For example, in a table such as an Adverse Events table, there are some standards that can be expected that allows us to constructing a basic `tfrmt`:

First, we know that the cells should be consistently formatted as follows, which allows us to create a body_plan:

>   - Total counts to be whole numbers 

Reviewer: ```suggestion
  - Total counts should be whole numbers 
```

>   - Cells under each treatment to be a combination of a count and percent of that population. When percent is 100 or zero, value should be blank.

Reviewer: ```suggestion
  - Cells within each treatment group column should be a combination of a count and percent of that population. When percent is 100 or zero, value should be blank.
```

>   - p-values to be formatted where: >.99 is displayed as ">0.99", <0.001 is displayed as "<0.001", and any values between should be displayed to three decimal places. Missing p-values should be replaced with "--".

Reviewer: ```suggestion
  - p-values should be formatted where: >.99 is displayed as ">0.99", <0.001 is displayed as "<0.001", and any values between should be displayed to three decimal places. Missing p-values should be replaced with "--".
```

> Next, we can apply standard to our organizations AE ARD format to ensure that names of are set consistently:

Reviewer: ```suggestion
Next, because our data is standardized as an analysis results dataset, we can pre-fill the expected column names. This also ensures that the data abides to the standard and names are set consistently:
```

  
```{r}

ae_tfrmt_template <- tfrmt(
    group = AEBODSYS,
    label = AETERM,
    param = param,
    column = treatment,
    value = value,
  body_plan = body_plan(
    ## All entries where the param column is `AEs` (representing total counts)
    frmt_structure(
      group_val = ".default",
      label_val = ".default",
      AEs = frmt("[XXX]")
      ),
    
    ## Combine entries where param column is `n` and `pct` to create a cell for
    ## that population
    frmt_structure(
      group_val = ".default",
      label_val = ".default",
      frmt_combine(
        "{n} {pct}",
        n = frmt("XXX"),
        pct = frmt_when(
          "==100" ~ "",
          "==0" ~ "",
          TRUE ~ frmt("(xx.x %)")
          )
        )
      ),
    
    ## All entries where param column is `pval`, format conditionally.
    ## When the value is missing, replace the NA with "--".
    frmt_structure(
       group_val = ".default", 
       label_val = ".default",
       pval = frmt_when(
         ">0.99" ~ ">0.99",
         "<0.001" ~ "<0.001",
         TRUE ~ frmt("x.xxx", missing = "--")
         )
    )
  )
)

```

## Piping - two layer tfrmts

> Now that we have a template to work off of, we can create a `tfrmt` specifically for the table we are creating for our study. This can be done a few ways. The first method that can be used is to pipe another tfrmt. The first argument of `trfmt` is "tfrmt_obj". This allows for `tfrmt` to be layered directly using a pipe. 

Reviewer: ```suggestion
Now we have a template to serve as the basis for our AE tables, which reduces much of the overhead in our formatting. We can continue to build on this template by making a study-specific `tfrmt` to be layered on top. This can be done a few ways. The first method that can be used is to pipe another tfrmt. The first argument of `trfmt` is "tfrmt_obj". This allows for `tfrmt` to be layered directly using a pipe. 
```

> We will create the new tfrmt with specific information that we know we have for this study, such as columns indicating order (ord1, ord2), we want to indent the row-groups as opposed to setting them into different columns, and how we want to align the columns

Reviewer: ```suggestion
We will create the new tfrmt with specific information that we know we have for this study, such as the columns indicating order (ord1, ord2), formatting of the row group labels, and column alignment.
```

```{r}

study_ae_tfrmt <- ae_tfrmt_template %>% 
  tfrmt(
    title = "Adverse Events for CDISC Pilot Study",
    subtitle = "Data subset to AEs with >10% prevalence in the High Dose group",
    
    ## Sorting columns of rows
    sorting_cols = c(ord1, ord2),
    
    ## Nest Preferred terms under SOC
    row_grp_plan = row_grp_plan(label_loc = element_row_grp_loc(location = "indented")),
    

>     ## alisgnment of columns

Reviewer: ```suggestion
    ## alignment of columns
```

      span_structure(
        "Xanomeline High Dose (N=84)",
        `n (%)` = `Xanomeline High Dose-n_pct` ,
        `[AEs]` = `Xanomeline High Dose-AEs`
      ),
      span_structure(
        "Xanomeline Low Dose (N=84)",
        `n (%)` = `Xanomeline Low Dose-n_pct` ,
        `[AEs]` = `Xanomeline Low Dose-AEs`
      ),
      span_structure("Placebo (N=86)",
                     `n (%)` = `Placebo-n_pct` ,
                     `[AEs]` = `Placebo-AEs`),
      span_structure(
        "Fisher's Exact p-values",
        `Placebo vs. Low Dose` = `fisher_pval-p_low` ,
        `Placebo vs. High Dose` = `fisher_pval-p_high`
      )
    )
  )

```

Now that we have the study specific `tfrmt`, we can apply it to our data:

```{r}

## filter to keep only AEs with >10% prevalence in the High Dose group
ae_data2 <- ae_data %>% 
  group_by(AEBODSYS, AETERM) %>% 
  mutate(pct_high = value[col2=="Xanomeline High Dose" & param=="pct"]) %>% 
  ungroup %>% 
  filter(pct_high >10) %>% 
  select(-pct_high) %>% 
  unite("treatment", c(col2, col1), sep = "-")

study_ae_tfrmt %>% 
  print_to_gt(ae_data2) %>% 
  tab_options(
    container.width = 1000
  )

```

## Multi-layer Example

> The previous example of a template tfrmt is very powerful in that it allows components to be reused across multiple projects and share the learning from prior work. However, the downside is that now this must be the "top level" `tfrmt`, as `tfrmt`'s may be layered off of this template, this must be at the beginning of the layering. 

Reviewer: Maybe, 
However, the downside is that we are limited to 2 layers. Our template `tfrmt` must be specified first, followed by our study-specific format.

> The previous example of a template tfrmt is very powerful in that it allows components to be reused across multiple projects and share the learning from prior work. However, the downside is that now this must be the "top level" `tfrmt`, as `tfrmt`'s may be layered off of this template, this must be at the beginning of the layering. 

Author: ```suggestion
The previous example of a template tfrmt is very powerful in that it allows components to be reused across multiple projects and share the learning from prior work. However, the downside is that we are limited to 2 layers. Our template `tfrmt` must be specified first, followed by our study-specific format.
```

```{r setup}
library(tlang)
library(dplyr)
library(gt)
library(tidyr)
```

A core design element of `tlang` is to offer the ability to layer and build template table formats that can be shared, improved, and eventually customized for a specific table. The idea being that organizations can develop a standard `tfrmt` object that can be reused across many different tables, and make it so programmers can focus on setting up the data as opposed to formatting of the table. 

There are two ways that `tfrmt` objects may be layered. The preferred method is through the first argument of `tfrmt` - `tfrmt_obj`. The second, but perhaps more explicit method is using the `layer_tfrmt` function. We will go through their applications below.  

> ## Creating a template

Reviewer 2: I think we might have been a touch hasty in making this vingette. I think we are better off making a canned tfrmt. There is a lot of say about creating templates so I think there should be a separate vingette for that. Just so the purpose of this one is a bit more focused

> ## Creating a template

Reviewer 2: I think we might have been a touch hasty in making this vingette. I think we are better off making a canned tfrmt. There is a lot of say about creating templates so I think there should be a separate vingette for that. Just so the purpose of this one is a bit more focused

Reviewer: i like that idea- separating the use of existing templates (typical user) from the creation of reusable templates (advanced)
